binaries/daemon/src/listener/mod.rs

```diff
@@ -176,34 +179,46 @@ where
         }
     }
 
-    async fn run_inner(&mut self) -> eyre::Result<()> {
+    async fn run_inner<C: Connection>(&mut self, mut connection: C) -> eyre::Result<()> {
         loop {
-            // receive the next node message
-            let message = match self
-                .connection
-                .receive_message()
-                .await
-                .wrap_err("failed to receive DaemonRequest")
-            {
-                Ok(Some(m)) => m,
+            let mut next_message = connection.receive_message();
+            let message = loop {
+                let next_event = if let Some(events) = &mut self.subscribed_events {
+                    Box::pin(events.recv()).fuse()
+                } else {
+                    Fuse::terminated()
+                };
+                let event = match future::select(next_event, next_message).await {
+                    future::Either::Left((event, n)) => {
+                        next_message = n;
+                        event
+                    }
+                    future::Either::Right((message, _)) => break message,
+                };
+                if let Some(event) = event {
+                    self.queue.push_back(Box::new(Some(event)));
+                    self.handle_events().await?;
+                }
+            };
+
+            // TODO: wait for event queue and connection simultaneously
```

Comment:
Are we not waiting for event queue and connection simultaneously line 191?

Reply:
We do, I just forgot to remove the TODO as well. Thanks for the hint, fixed in 974d3156977e41de88659f8255db165db8f7224f